File: exercises/exercise-1-testing-java/src/main/java/de/emaarco/example/adapter/outbound/db/bike/BikeSubscriptionPersistenceAdapter.java
package de.emaarco.example.adapter.outbound.db.bike;

import de.emaarco.example.application.port.outbound.bike.BikeSubscriptionRepository;
import de.emaarco.example.domain.bike.BikeSubscription;
import de.emaarco.example.domain.bike.BikeSubscriptionId;
import org.springframework.stereotype.Component;

import java.util.NoSuchElementException;

@Component
public class BikeSubscriptionPersistenceAdapter implements BikeSubscriptionRepository {

    private final BikeSubscriptionJpaRepository jpaRepository;
    private final BikeSubscriptionEntityMapper mapper;

    public BikeSubscriptionPersistenceAdapter(
        BikeSubscriptionJpaRepository jpaRepository,
        BikeSubscriptionEntityMapper mapper
    ) {
        this.jpaRepository = jpaRepository;
        this.mapper = mapper;
    }

    @Override
    public BikeSubscription save(BikeSubscription subscription) {
        var entity = mapper.toEntity(subscription);
        var saved = jpaRepository.save(entity);
        return mapper.toDomain(saved);
    }

    @Override
    public BikeSubscription find(BikeSubscriptionId id) {
        var entity = jpaRepository.findById(id.value())
            .orElseThrow(() -> new NoSuchElementException("BikeSubscription with id " + id.value() + " not found"));
        return mapper.toDomain(entity);
    }

    @Override
    public BikeSubscription search(BikeSubscriptionId id) {
        return jpaRepository.findById(id.value())
            .map(mapper::toDomain)
            .orElse(null);
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/io/miragon/example/adapter/outbound/db/bike/BikeSubscriptionPersistenceAdapter.kt
package io.miragon.example.adapter.outbound.db.bike

import io.miragon.example.application.port.outbound.bike.BikeSubscriptionRepository
import io.miragon.example.domain.bike.BikeSubscription
import io.miragon.example.domain.bike.BikeSubscriptionId
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Component

@Component
class BikeSubscriptionPersistenceAdapter(
    private val jpaRepository: BikeSubscriptionJpaRepository,
    private val mapper: BikeSubscriptionEntityMapper
) : BikeSubscriptionRepository {

    override fun save(subscription: BikeSubscription): BikeSubscription {
        val entity = mapper.toEntity(subscription)
        val saved = jpaRepository.save(entity)
        return mapper.toDomain(saved)
    }

    override fun find(id: BikeSubscriptionId): BikeSubscription {
        val entity = jpaRepository.findByIdOrNull(id.value)
            ?: throw NoSuchElementException("BikeSubscription with id ${id.value} not found")
        return mapper.toDomain(entity)
    }

    override fun search(id: BikeSubscriptionId): BikeSubscription? {
        return jpaRepository.findByIdOrNull(id.value)?.let { mapper.toDomain(it) }
    }
}
